File: JavaScript/reverseString.js

function reverseString(inputstring) {
    let  i =0;
    let outputString = '';
    let outputString1 = '';
    //let outputString2;

    /* method1: use of for loop*/
    for (i=inputstring.length-1; i>=0; i--){
        //outputString.push(inputstring[i]);
        outputString += inputstring[i];

    }
    console.log(outputString);

    /*Method 2: using split, reverse and join methods  */
    const arr = inputstring.split('');
    arr.reverse();
    arr.join('');
    console.log(arr);

    //Method 3: using for loop for a string 
    
    for (let output of inputstring) {   
        outputString1 = output+outputString1;
    }
    console.log(outputString1);

    //Method 4: 
    let arr1 = inputstring.split('');
    
    let stringFinalOutput = arr1.reduce( 
        (el,outputString2) => {
            return outputString2+el;
        } ,'');
    console.log(stringFinalOutput);

}

//module.exports =  reverseString;

reverseString('Hello World');

File: Python/reverse_string.py
def reverse_string(str):    
    """ Method 1:  using for loop """
    output_str=''
    for i  in  str:
        output_str =  i + output_str
    #print(output_str)
    
    """ Method 2: converting string to list """
    output_str1=''.join(list(str[::-1]))
    """ the above line roughly translates to the below:
    output_str1 = list(str)
    finalOutputStr = ''.join(output_str1[::-1]) 
    """
    #print(output_str1)

    """ Method 3: using  reversed function. reversed() function  is returns
    a reversed iterator. join function is used to join back strings using empty separator """
    output_str2=''.join(reversed(str))
    #print(output_str2)
    """ Method 4: using recursive functions """
    
    def recursive_function(str1=str):

        if len(str1) ==  0:
            return str1
        else:
             
            return  recursive_function(str1[1:]) + str1[0]
    output = recursive_function()
    print(output)
#print(reverse_string('hello world'))

def main():
    reverse_string('hello world')

if __name__ == '__main__':
    main()

